### DISCUSSION TITLE
Implementing evil-commentary's gy map to comment and yank

### DISCUSSION THREAD

**Discussion by cherryman** [nvim-mini/mini.nvim#1025] [Ideas]
Was wondering if it's possible to implement this using `mini.comment` or even with the now built-in comment mapping. I'm unfortunately not familiar enough with vimscript to figure out how to cleanly do this.

I've implemented this in [cherryman/vim-commentary-yank](https://github.com/cherryman/vim-commentary-yank/), but looking for a neovim specific solution.

**Comment by echasnovski** [comment]
Not *exactly* what you are looking for, but you can do anything inside [`config.hooks.pre`](https://github.com/echasnovski/mini.nvim/blob/12d275f4c3aeddfa12a758a69aff65090613ca34/doc/mini-comment.txt#L110). Here is [an example of always yanking commented text](https://github.com/echasnovski/mini.nvim/issues/701#issuecomment-1948501258).

I am not sure how to properly achieve similar with built-in commenting. In Visual mode it can be done with `:xmap gy ygvgc` (`y` - yank visual selection, `gv` - restore visual selection, `gc` - comment visual selection). Note that it needs `remap = true` to achieve this with Lua's `vim.keymap.set()`.

**Comment by cherryman** [comment]
For anyone else who runs into this, if you're willing to rely on implementation details of the [original PR](https://github.com/neovim/neovim/pull/28176/files), then this works:

```lua
_G.__commentyank = function(mode)
  if mode == nil then
    vim.o.operatorfunc = 'v:lua.__commentyank'
    return 'g@'
  end

  local mark_from, mark_to = "'[", "']"
  local lnum_from, col_from = vim.fn.line(mark_from), vim.fn.col(mark_from)
  local lnum_to, col_to = vim.fn.line(mark_to), vim.fn.col(mark_to)

  if (lnum_from > lnum_to) or (lnum_from == lnum_to and col_from > col_to) then
    return
  end

  local l = vim.api.nvim_buf_get_lines(0, lnum_from - 1, lnum_to, false)
  vim.fn.setreg(vim.v.register, l)

  -- WARN: Relying on implementation details of `gc`.
  require('vim._comment').toggle_lines(lnum_from, lnum_to, vim.api.nvim_win_get_cursor(0))

  return ''
end
```

Then it's enough to map `gy` to `_G.__commentyank` with `expr = true`.
